flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java

```diff
@@ -158,7 +158,16 @@ protected void processInput(MailboxDefaultAction.Controller controller) throws E
         // compatibility reasons with the current source interface (source functions run as a loop,
         // not in steps).
         sourceThread.setTaskDescription(getName());
-        sourceThread.start();
+
+        if (operatorChain.isFinishedOnRestore()) {
+            LOG.debug(
+                    "Legacy source {} skip execution since the task is finished on restore",
+                    getTaskNameWithSubtaskAndId());
+            sourceThread.getCompletionFuture().complete(null);
```

Comment:
Hi Dawid~ I think previously the `mailbox.suspend` is still done after the future completed since it is register to the future with `whenComplete` ~? Perhaps we might have some code duplicate if we directly suspend the mailbox for the finished case since here we would still need to complete the future, which would already trigger the `whenComplete` stage and suspend the mailbox with the current implementation~?

Reply:
Yes, you are right!